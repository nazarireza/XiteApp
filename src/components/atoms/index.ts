export * from './IconClose';
export * from './IconSearch';
export * from './IconBack';
export * from './Space';
export * from './IconButton';
export * from './IconFilter';
export * from './GenreItem';
export * from './VideoItem';
