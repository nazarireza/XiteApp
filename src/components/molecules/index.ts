export * from './SearchBox';
export * from './GenresList';
export * from './VideosList';
export * from './Header';
export * from './GenresWrapList';
export * from './YearsWrapList';
export * from './VideoCompactList';
