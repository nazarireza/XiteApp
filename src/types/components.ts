export type IconBaseProps = {
  color?: string;
  size?: number;
};

export type onListItemSelect<T> = (item: {
  item: T;
  index: number;
  selected?: boolean;
}) => void;
