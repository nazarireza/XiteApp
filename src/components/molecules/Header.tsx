import React, {memo} from 'react';
import {StyleSheet, View, TouchableOpacity, Text} from 'react-native';
import {colors, dictionary, typography} from '../../assets';
import {IconBack, Space} from '../atoms';

interface HeaderProps {
  title?: string;
  onBackPress?: () => void;
  actionTitle?: string;
  onActionPress?: () => void;
  dissmissable?: boolean;
}

export const Header: React.FC<HeaderProps> = memo(
  ({
    title = dictionary.filterModalTitle,
    actionTitle,
    onActionPress,
    onBackPress,
    dissmissable = true,
  }) => {
    const hasAction = !!actionTitle;
    return (
      <View style={styles.container}>
        <TouchableOpacity
          disabled={!dissmissable}
          style={styles.titleContainer}
          activeOpacity={0.7}
          onPress={onBackPress}>
          {dissmissable && (
            <>
              <IconBack color={colors.text.primary} />
              <Space size={8} />
            </>
          )}
          <Text style={typography.header}>{title}</Text>
        </TouchableOpacity>
        {hasAction && (
          <TouchableOpacity
            activeOpacity={0.7}
            style={styles.actionContainer}
            onPress={onActionPress}>
            <Text style={typography.headerAction}>{actionTitle}</Text>
          </TouchableOpacity>
        )}
      </View>
    );
  },
);

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    alignItems: 'center',
    paddingVertical: 8,
  },
  titleContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    flex: 1,
  },
  actionContainer: {
    alignItems: 'center',
    justifyContent: 'center',
  },
});
