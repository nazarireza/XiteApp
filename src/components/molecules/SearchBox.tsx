import React, {forwardRef, memo} from 'react';
import {StyleSheet, View, TextInput, TextInputProps} from 'react-native';
import {colors, dictionary} from '../../assets';
import {IconButton, IconClose, IconSearch, Space} from '../atoms';

interface SearchBoxProps {
  value?: string;
  placeHolder?: string;
  onChange?: (keyword: string) => void;
}

export const SearchBox = memo(
  forwardRef<TextInput, TextInputProps & SearchBoxProps>(
    (
      {placeHolder = dictionary.searchBoxPlaceHolder, onChange, value, ...rest},
      ref,
    ) => {
      const hasValue = !!value;
      return (
        <View style={styles.container}>
          <IconSearch color={colors.text.secondary} />
          <Space size={8} />
          <TextInput
            ref={ref}
            onChangeText={onChange}
            value={value}
            style={styles.input}
            placeholder={placeHolder}
            placeholderTextColor={colors.text.secondary}
            returnKeyType="search"
            {...rest}
          />
          {hasValue && (
            <IconButton onPress={() => onChange?.('')}>
              <IconClose />
            </IconButton>
          )}
        </View>
      );
    },
  ),
);

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    alignItems: 'center',
    backgroundColor: colors.background.card,
    paddingHorizontal: 16,
    borderRadius: 30,
  },
  input: {
    flex: 1,
    color: colors.text.primary,
    paddingVertical:12,
  },
});
