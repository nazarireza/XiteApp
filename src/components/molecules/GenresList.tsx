import React, {memo, useCallback} from 'react';
import {
  StyleSheet,
  FlatListProps,
  FlatList,
  ListRenderItem,
  View,
  TouchableOpacityProps,
} from 'react-native';
import {colors} from '../../assets';
import {Genre} from '../../types';
import {IconButton, IconFilter, Space, GenreItem, onListItemSelect} from '../atoms';

interface GenresListProps {
  onFilterPress?: () => void;
  selectedItems?: Array<Genre>;
  onItemSelect?: onListItemSelect<Genre>;
}

export const GenresList: React.FC<FlatListProps<Genre> & GenresListProps> =
  memo(({onFilterPress, selectedItems, onItemSelect, ...rest}) => {
    const renderItem: ListRenderItem<Genre> = useCallback(
      ({item, index}) => {
        const selected = selectedItems?.some(p => p.id == item.id);
        return (
          <GenreItem
            {...item}
            selected={selected}
            onPress={() => onItemSelect?.({item, index, selected: !selected})}
          />
        );
      },
      [selectedItems],
    );

    return (
      <View style={styles.container}>
        <FlatList
          ListHeaderComponent={() => <FilterButton onPress={onFilterPress} />}
          ItemSeparatorComponent={() => <Space size={8} />}
          showsHorizontalScrollIndicator={false}
          horizontal
          contentContainerStyle={styles.scrollContainer}
          {...rest}
          renderItem={renderItem}
        />
      </View>
    );
  });

const FilterButton: React.FC<TouchableOpacityProps> = memo(props => {
  return (
    <IconButton style={styles.filterButton} {...props}>
      <IconFilter />
    </IconButton>
  );
});

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
  },
  scrollContainer: {
    paddingHorizontal: 24,
    alignItems: 'center',
  },
  filterButton: {
    marginEnd: 8,
    backgroundColor: colors.primary,
    paddingVertical: 4,
    paddingHorizontal: 16,
    borderRadius: 5,
  },
});
