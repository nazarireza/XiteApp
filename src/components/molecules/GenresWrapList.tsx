import React, {memo, useCallback} from 'react';
import {Genre} from '../../types';
import {
  ToggleButton,
  WrapList,
  WrapListProps,
  WrapListRenderItem,
} from '../atoms';

export const GenresWrapList: React.FC<WrapListProps<Genre>> = memo(props => {
  const renderItem: WrapListRenderItem<Genre> = useCallback(
    ({item, index, selected, onSelect}) => {
      return (
        <ToggleButton
          title={item.name}
          selected={selected}
          onPress={() => onSelect?.({item, index, selected: !selected})}
        />
      );
    },
    [],
  );

  return (
    <WrapList
      {...props}
      {...{renderItem}}
      keyExtractor={(item: Genre) => `${item.id}`}
    />
  );
});
