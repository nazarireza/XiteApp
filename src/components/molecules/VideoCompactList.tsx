import React, {memo, useCallback} from 'react';
import {
  StyleSheet,
  FlatListProps,
  FlatList,
  ListRenderItem,
} from 'react-native';
import {NormalizedVideo} from '../../types';
import {Space, onListItemSelect, VideoCompactItem} from '../atoms';

interface VideosCompactListProps {
  onItemSelect?: onListItemSelect<NormalizedVideo>;
}

export const VideosCompactList: React.FC<
  FlatListProps<NormalizedVideo> & VideosCompactListProps
> = memo(({onItemSelect, ...rest}) => {
  const renderItem: ListRenderItem<NormalizedVideo> = useCallback(
    ({item, index}) => {
      return (
        <VideoCompactItem
          {...item}
          onPress={() => onItemSelect?.({item, index})}
        />
      );
    },
    [],
  );

  return (
    <FlatList
      keyExtractor={item => `${item.id}`}
      ItemSeparatorComponent={() => <Space />}
      contentContainerStyle={styles.scrollContainer}
      {...rest}
      renderItem={renderItem}
    />
  );
});

const styles = StyleSheet.create({
  scrollContainer: {},
});
