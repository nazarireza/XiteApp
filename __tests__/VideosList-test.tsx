import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import {VideosList} from '../src/components/molecules';

const data = [
  {
    id: 501437,
    artist: 'Pants Velour',
    title: 'All In',
    release_year: 2014,
    genre_id: 14,
    image_url:
      'https://raw.githubusercontent.com/XiteTV/frontend-coding-exercise/679a82b1e7110c16e14412f1debaa118c10078a9/images/501437/images/app/w522_h292.jpg',
  },
  {
    id: 501649,
    artist: 'El Koala',
    title: 'Veni paca to',
    release_year: 2014,
    genre_id: 8,
    image_url:
      'https://raw.githubusercontent.com/XiteTV/frontend-coding-exercise/679a82b1e7110c16e14412f1debaa118c10078a9/images/501649/images/app/w522_h292.jpg',
  },
  {
    id: 501895,
    artist: 'Tom Petty and the Heartbreakers',
    title: 'I Should Have Known It',
    release_year: 2010,
    genre_id: 8,
    image_url:
      'https://raw.githubusercontent.com/XiteTV/frontend-coding-exercise/679a82b1e7110c16e14412f1debaa118c10078a9/images/501895/images/app/w522_h292.jpg',
  },
];

it('renders correctly', () => {
  const wrapper = renderer.create(<VideosList data={data} />).toJSON();
  expect(wrapper).toMatchSnapshot();
});
