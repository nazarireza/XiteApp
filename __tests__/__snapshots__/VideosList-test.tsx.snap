// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<RCTScrollView
  ItemSeparatorComponent={[Function]}
  contentContainerStyle={
    Object {
      "paddingHorizontal": 4,
      "paddingVertical": 16,
    }
  }
  data={
    Array [
      Object {
        "artist": "Pants Velour",
        "genre_id": 14,
        "id": 501437,
        "image_url": "https://raw.githubusercontent.com/XiteTV/frontend-coding-exercise/679a82b1e7110c16e14412f1debaa118c10078a9/images/501437/images/app/w522_h292.jpg",
        "release_year": 2014,
        "title": "All In",
      },
      Object {
        "artist": "El Koala",
        "genre_id": 8,
        "id": 501649,
        "image_url": "https://raw.githubusercontent.com/XiteTV/frontend-coding-exercise/679a82b1e7110c16e14412f1debaa118c10078a9/images/501649/images/app/w522_h292.jpg",
        "release_year": 2014,
        "title": "Veni paca to",
      },
      Object {
        "artist": "Tom Petty and the Heartbreakers",
        "genre_id": 8,
        "id": 501895,
        "image_url": "https://raw.githubusercontent.com/XiteTV/frontend-coding-exercise/679a82b1e7110c16e14412f1debaa118c10078a9/images/501895/images/app/w522_h292.jpg",
        "release_year": 2010,
        "title": "I Should Have Known It",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  viewabilityConfigCallbackPairs={Array []}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "flexDirection": "row",
          }
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          nativeID="animatedComponent"
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "marginEnd": 8,
              "opacity": 1,
              "width": 367,
            }
          }
        >
          <Image
            source={
              Object {
                "uri": "https://raw.githubusercontent.com/XiteTV/frontend-coding-exercise/679a82b1e7110c16e14412f1debaa118c10078a9/images/501437/images/app/w522_h292.jpg",
              }
            }
            style={
              Object {
                "height": 200,
              }
            }
          />
          <View
            style={
              Object {
                "height": 8,
                "width": 8,
              }
            }
          />
          <View
            style={
              Object {
                "paddingHorizontal": 8,
              }
            }
          >
            <Text
              adjustsFontSizeToFit={true}
              numberOfLines={1}
              style={
                Object {
                  "color": "#F1F2F4",
                  "fontFamily": "Montserrat",
                  "fontSize": 15,
                }
              }
            >
              All In
            </Text>
            <Text
              adjustsFontSizeToFit={true}
              numberOfLines={1}
              style={
                Object {
                  "color": "#797D95",
                  "fontFamily": "Montserrat",
                  "fontSize": 13,
                }
              }
            >
              Pants Velour
            </Text>
          </View>
        </View>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          nativeID="animatedComponent"
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "marginEnd": 8,
              "opacity": 1,
              "width": 367,
            }
          }
        >
          <Image
            source={
              Object {
                "uri": "https://raw.githubusercontent.com/XiteTV/frontend-coding-exercise/679a82b1e7110c16e14412f1debaa118c10078a9/images/501649/images/app/w522_h292.jpg",
              }
            }
            style={
              Object {
                "height": 200,
              }
            }
          />
          <View
            style={
              Object {
                "height": 8,
                "width": 8,
              }
            }
          />
          <View
            style={
              Object {
                "paddingHorizontal": 8,
              }
            }
          >
            <Text
              adjustsFontSizeToFit={true}
              numberOfLines={1}
              style={
                Object {
                  "color": "#F1F2F4",
                  "fontFamily": "Montserrat",
                  "fontSize": 15,
                }
              }
            >
              Veni paca to
            </Text>
            <Text
              adjustsFontSizeToFit={true}
              numberOfLines={1}
              style={
                Object {
                  "color": "#797D95",
                  "fontFamily": "Montserrat",
                  "fontSize": 13,
                }
              }
            >
              El Koala
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "height": 30,
            "width": 30,
          }
        }
      />
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "flexDirection": "row",
          }
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          nativeID="animatedComponent"
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "marginEnd": 8,
              "opacity": 1,
              "width": 367,
            }
          }
        >
          <Image
            source={
              Object {
                "uri": "https://raw.githubusercontent.com/XiteTV/frontend-coding-exercise/679a82b1e7110c16e14412f1debaa118c10078a9/images/501895/images/app/w522_h292.jpg",
              }
            }
            style={
              Object {
                "height": 200,
              }
            }
          />
          <View
            style={
              Object {
                "height": 8,
                "width": 8,
              }
            }
          />
          <View
            style={
              Object {
                "paddingHorizontal": 8,
              }
            }
          >
            <Text
              adjustsFontSizeToFit={true}
              numberOfLines={1}
              style={
                Object {
                  "color": "#F1F2F4",
                  "fontFamily": "Montserrat",
                  "fontSize": 15,
                }
              }
            >
              I Should Have Known It
            </Text>
            <Text
              adjustsFontSizeToFit={true}
              numberOfLines={1}
              style={
                Object {
                  "color": "#797D95",
                  "fontFamily": "Montserrat",
                  "fontSize": 13,
                }
              }
            >
              Tom Petty and the Heartbreakers
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
